contracts/combat/B5_Combat.cairo

```diff
@@ -0,0 +1,393 @@
+from starkware.cairo.common.pow import pow
+
+# a min delay between attacks on a Realm; it can't
+# be attacked again during cooldown
+const ATTACK_COOLDOWN_PERIOD = 86400  # 1 day
+
+const COMBAT_TYPE_ATTACK_VS_DEFENSE = 1
+const COMBAT_TYPE_WISDOM_VS_AGILITY = 2
+const COMBAT_OUTCOME_ATTACKER_WINS = 1
+const COMBAT_OUTCOME_DEFENDER_WINS = 2
+
+# namespace constants deliberately start at 1 to
+# 1) translate in a straighforward way to "human, 1-based index" land
+# 2) to differentiate between uninitialized value (i.e. 0 in Cairo)
+
+namespace TroopId:
+    const Watchman = 1
+    const Guard = 2
+    const GuardCaptain = 3
+    const Squire = 4
+    const Knight = 5
+    const KnightCommander = 6
+    const Scout = 7
+    const Archer = 8
+    const Sniper = 9
+    const Scorpio = 10
+    const Baillista = 11
+    const Catapult = 12
+    const Apprentice = 13
+    const Mage = 14
+    const Arcanist = 15
+    const GrandMarshal = 16
+end
+
+namespace TroopType:
+    const Melee = 1
+    const Ranged = 2
+    const Siege = 3
+end
+
+struct Troop:
+    member type : felt  # TroopType
+    member tier : felt
+    member agility : felt
+    member attack : felt
+    member defense : felt
+    member vitality : felt
+    member wisdom : felt
+end
+
+# TODO: maybe use vector from 0xnoncents to build the full squad?
+# https://gist.github.com/0xNonCents/11f8488cd800fec49bf4cd89495722b4
+struct Squad:
+    # tier 1 troops
+    member t1_1 : Troop
+    member t1_2 : Troop
+    member t1_3 : Troop
+    member t1_4 : Troop
+    member t1_5 : Troop
+    member t1_6 : Troop
+    member t1_7 : Troop
+    member t1_8 : Troop
+    member t1_9 : Troop
+    member t1_10 : Troop
+    member t1_11 : Troop
+    member t1_12 : Troop
+    member t1_13 : Troop
+    member t1_14 : Troop
+    member t1_15 : Troop
+    member t1_16 : Troop
+
+    # tier 2 troops
+    member t2_1 : Troop
+    member t2_2 : Troop
+    member t2_3 : Troop
+    member t2_4 : Troop
+    member t2_5 : Troop
+    member t2_6 : Troop
+    member t2_7 : Troop
+    member t2_8 : Troop
+
+    # tier 3 troop
+    member t3_1 : Troop
+end
+
+struct PackedSquad:
+    # one packed troop fits into 7 bytes
+    # one felt is ~31 bytes -> can hold 4 troops
+    # a squad has 25 troops -> fits into 7 felts when packed
+    member p1 : felt  # packed Troops t1_1 ... t1_4
+    member p2 : felt  # packed Troops t1_5 ... t1_8
+    member p3 : felt  # packed Troops t1_9 ... t1_12
+    member p4 : felt  # packed Troops t1_13 ... t1_16
+    member p5 : felt  # packed Troops t2_1 ... t2_4
+    member p6 : felt  # packed Troops t2_5 ... t2_8
+    member p7 : felt  # packed Troop t3_1
+end
+
+struct SquadStats:
+    member agility : felt
+    member attack : felt
+    member defense : felt
+    member vitality : felt
+    member wisdom : felt
+end
+
+const SHIFT = 0x100  # used for packing
+
+@view
+func get_troop{range_check_ptr, bitwise_ptr : BitwiseBuiltin*}(troop_id : felt) -> (t : Troop):
```

Comment:
We could create specific logic contract for these state functions. 

Eg: get_troop, get_kraken

Then we can make them an upgradeable proxy for easily increasing the enemies and also keeping them all in the same place

Reply:
Agreed. There's a lot of optimizations and cleanups we should do during / after merging.

On that point, I was trying to do this more generic, but ran into Cairo limitations - can't pass a pointer to a type (like a Troop struct) or build a struct in a generic way :( Shame, would be really useful.